### PR TITLE
[pkcs12] add fingerprint support for unparseable x509 certs

### PR DESCRIPTION
**Pull request by quinlanj**
# why
fixes https://github.com/expo/expo-cli/issues/2845

`node-forge` only supports x509 certificates that is RSA encrypted. This is causing problems because some of our users are passing in DSA encrypted x509s. 

Fortunately, if `node-forge` errors when parsing the x509 certificate, it'll pass back its asn1 representation instead (see [here](https://github.com/digitalbazaar/forge/blob/1887cfce43a8f5ca9cb5c256168cf12ce1715ecf/lib/pkcs12.js#L679)). We can use this asn1 value to compute the certificate fingerprints instead.

This PR does the following:
- adds `getX509Asn1ByFriendlyName`: extract the asn1 representation of the x509 instead of trying to parse it into a certificate object
- adds `getAsn1Hash`: computes the hash of an asn1 object
- adds more tests, re-organizes current tests

### CODE REVIEW

File: packages/pkcs12/src/index.ts [hunk 1]
```diff
@@ -43,11 +39,40 @@ export function getX509CertificateByFriendlyName(
   if (!bags || bags.length === 0) {
     return null;
   }
-  const certificate = bags[0].cert;
-  if (!certificate) {
+  return extractx509CertificateFromBag(bags[0]);
+}
+
+function extractx509CertificateFromBag(bag: forge.pkcs12.Bag): forge.pki.Certificate {
```
Comment (ide) [inline packages/pkcs12/src/index.ts:45]:
```suggestion
function extractX509CertificateFromBag(bag: forge.pkcs12.Bag): forge.pki.Certificate {
```

File: packages/pkcs12/src/index.ts [hunk 2]
```diff
@@ -43,11 +39,40 @@ export function getX509CertificateByFriendlyName(
...
+  const { cert, asn1 } = bag;
+  if (!cert && asn1) {
+    // if asn1 is present but certificate isnt, the certificate type was unknown
+    // github.com/digitalbazaar/forge/blob/1887cfce43a8f5ca9cb5c256168cf12ce1715ecf/lib/pkcs12.js#L703
+    throw new Error('PKCS12: unknown x509 certificate type');
```
Comment (ide) [inline packages/pkcs12/src/index.ts:50]:
```suggestion
    throw new Error('PKCS12: unknown X.509 certificate type');
```

File: packages/pkcs12/src/__tests__/index-test.ts [hunk 3]
```diff
@@ -39,7 +41,29 @@ const uppercaseAliasKeystoreP12 = {
...
+/**
+ * Generate DSA certificate:
+ * openssl dsaparam -out params.pem 3072
+ * openssl gendsa -out key.pem params.pem
+ * openssl req -new -key key.pem -out req.pem
+ * openssl x509 -req -in req.pem -signkey key.pem -out certificate.cer
+ *
+ * Package into a PKCS#12 keystore:
+ * keytool -import -alias test -file certificate.cer -keystore dsapkcs.keystore -storetype pkcs12
+ *
```
Comment (ide) [inline packages/pkcs12/src/__tests__/index-test.ts:54]: How are the fingerprints calculated?

File: packages/pkcs12/src/__tests__/index-test.ts [hunk 4]
```diff
@@ -87,6 +111,79 @@ describe('reading PKCS#12 files', () => {
     expect(sha1Fingerprint).toEqual(expectedSha1Fingerprint);
     expect(sha256Fingerprint).toEqual(expectedSha256Fingerprint);
   });
+});
+describe('computing fingerprints from the asn1 representation of a x509 certificate', () => {
+  it('computes fingerprints of unsupported certificates using their asn1 values in p12 keystores', async () => {
```
Comment (ide) [inline packages/pkcs12/src/__tests__/index-test.ts:116]:
```suggestion
  it.each([
    ['supported RSA certificates', dsaKeystoreP12],
    ['unsupported RSA certificates', keystoreP12],
  ])('computes fingerprints of %s using their ASN.1 values in PKCS #12 keystores', async (_description, keystoreP12) => {
```

**Review by ide — CHANGES_REQUESTED** [review]

File: packages/pkcs12/src/__tests__/index-test.ts (hunk 3 above)
Comment (quinlanj) [reply packages/pkcs12/src/__tests__/index-test.ts:54]: fingerprints are calculated by getting the asn.1 representation of the certificate, then hashing it.

**Review by ide — APPROVED** [review]

**Pull request merged**